streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java

```diff
@@ -1500,68 +1516,71 @@ private void validateRackAwarenessConfiguration() {
-                        if (!clientProvidedProps.get(config).equals(CONSUMER_EOS_OVERRIDES.get(config))) {
-                            log.warn(String.format(nonConfigurableConfigMessage,
-                                    "consumer", config, eosMessage, clientProvidedProps.get(config), CONSUMER_EOS_OVERRIDES.get(config)));
-                            clientProvidedProps.remove(config);
-                        }
-                    } else if (PRODUCER_EOS_OVERRIDES.containsKey(config)) {
-                        if (!clientProvidedProps.get(config).equals(PRODUCER_EOS_OVERRIDES.get(config))) {
-                            log.warn(String.format(nonConfigurableConfigMessage,
-                                    "producer", config, eosMessage, clientProvidedProps.get(config), PRODUCER_EOS_OVERRIDES.get(config)));
-                            clientProvidedProps.remove(config);
-                        }
-                    } else if (ProducerConfig.TRANSACTIONAL_ID_CONFIG.equals(config)) {
-                        log.warn(String.format(nonConfigurableConfigMessage,
-                            "producer", config, eosMessage, clientProvidedProps.get(config), "<appId>-<generatedSuffix>"));
-                        clientProvidedProps.remove(config);
-                    }
-                }
+    private void validateConsumerPropertyMap(final Map<String, Object> props) {
+        if (eosEnabled) {
+            // Iterate over KS_CONTROLLED_CONSUMER_CONFIGS_EOS_ENABLED and override values if set
+            for (final Map.Entry<String, Object> entry : KS_CONTROLLED_CONSUMER_CONFIGS_EOS_ENABLED.entrySet()) {
+                overwritePropertyMap(props, entry.getKey(), entry.getValue(), "consumer");
+            }
+            verifyMaxInFlightRequestPerConnection(props.get(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION));
+
+        } else {
+            // Iterate over KS_CONTROLLED_CONSUMER_CONFIGS and override values if set
+            for (final Map.Entry<String, Object> entry : KS_CONTROLLED_CONSUMER_CONFIGS.entrySet()) {
+                overwritePropertyMap(props, entry.getKey(), entry.getValue(), "consumer");
             }
         }
+    }
 
+    private void validateProducerPropertyMap(final Map<String, Object> props) {
         if (eosEnabled) {
-            verifyMaxInFlightRequestPerConnection(clientProvidedProps.get(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION));
+            // Iterate over KS_CONTROLLED_CONSUMER_CONFIGS_EOS_ENABLED and override values if set
+            for (final Map.Entry<String, Object> entry : KS_CONTROLLED_PRODUCER_CONFIGS_EOS_ENABLED.entrySet()) {
+                overwritePropertyMap(props, entry.getKey(), entry.getValue(), "producer");
+            }
+            verifyMaxInFlightRequestPerConnection(props.get(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION));
+
+        }
+
+        if (props.containsKey(ProducerConfig.PARTITIONER_CLASS_CONFIG)) {
```

Comment:
I am actually not sure from top of my head what @ableegoldman meant by "it would break Kafka Streams" is set?

Would be good to clarify if we actually need to do anything about it or not.

Reply:
Thanks Sophie. One small correction: it won't be used for changelogs, because for changelogs KS sets the partition  number expliclity base on the "task index" and thus neither the producer `Partioner` not a `StreamsPartitioner` is called. It's a hard-code 1:1 mapping (luckily).

Anyway...

I have no problem with disallowing the config in the future, however, it would require a KIP IMHO. As you pointed out, there might simple read-filter/map-write apps which could actually use a range-partitioner (as an example), and we could break them.

Thus, I would recommend that this PR does no do anything with regard to `partitioner.class` config, and if we feel we should disallow setting it going forward, we file a Jira and somebody can pick it up an do a proper KIP for it.

Reply:
Sounds good to me

Reply:
This was largely based on experience since I've seen a few users attempt to set the producer's partitioner because they simply were not aware of the Streams-specific `StreamPartitioner`. So while we don't need to prohibit the ProducerConfig from being used with Streams, I would suggest we still log a warning if the config is set and point users to the StreamPartitioner. Something like:

"To customize the partitioning within Kafka Streams, It is recommended to implement a StreamPartitioner instead of configuring a regular Partitioner with the partitioner.class config"